### ISSUE TITLE
Allow evaluation of constant locals and variables prior to module configuration

### ISSUE THREAD

**Issue by cam72cam** [opentofu/opentofu#1042]
### Summary

Many users expect to be able to DRY up their large configurations by using locals/variables in module sources and backend configurations (among other places).  This is currently prohibited and users are forced to do a lot of manual work or use additional tooling to work around this limitation.  We should be able to allow this functionality, with some caveats and guardrails.

### Problem Statement

One of the more frustrating features with the OpenTofu configuration language is the inability to use variables and locals in many common areas.  Two of the most requested being module sources and backed configurations.  This limitation exists as many parts of the configuration can only be evaluated during the plan and apply stages.  However, certain variables and locals *can* be evaluated in a static fashion prior to plan/apply and should be able to be used in many currently restricted areas.  This represents a massive improvement for many common use cases.

## Module Sources
The most common use case I have seen requested is the ability to specify module versions as a local variable.  Without the ability to modify the source attribute from a constant expression, bumping the version of org/my-utils is a tedious and error prone process.
```hcl
locals {
    util_version = 1.5.2
}

# Consider 20-30 similar instances of this module in half a dozen files
module {
    source = "git@github.com:org/my-utils?ref=${local.util_version}"
}
```

## Backend Configurations
Additionally users frequently request the ability to configure backends via locals/variables
```hcl
locals {
    region = "us-east-1"
}

terraform {
    backend "s3" {
        region = locals.region
    }
    required_providers {
        "aws_primary" = {
            source = "hashicorp/aws"
            region = locals.region
        }
    }
}
```

There are additional use cases, such as lifecycles, that are mentioned in the referenced issues and could be an extension off of this initial work.

## Pass-through Constant Variables
Some users have a legitimate use case for also allowing for some static fields to be passed in via variables.  A common use case is switching to a development version of a given module.

### main.tf
```hcl
locals {
    util_dev = true
}

module "submod" {
    source = "./submod/"
    util_dev = local.util_dev
}
```
### ./submod/main.tf
```hcl
variable "util_dev" {
    type = "bool"
}

locals {
    util_version = "1.5.2${var.util_dev ? "-dev" : ""}"
}

module {
    source = "git@github.com:org/my-utils?ref=${local.util_version}"
}
```

## ~~Label Interpolation (out-of-scope, technical discovery only)~~

~~Additionally, some users requested the ability to parameterize block labels.  This is possible, but I'm not sure we want to tackle it as part of this RFC.~~
```hcl
locals {
   prefix  = "org-"
}

module "{local.prefix}-mod" {
  source = "./src"
}
```
~~Note that the `$` is omitted from the interpolation as it is forbidden in the current HCL parser.~~

### Technical Description

The current workflow for modules is:
* For each path, create a ModuleFile:
  - Read the file contents
  - Parse the contents into a generic hcl structure
  - For each block:
    - Validate the block structure
    - Add a representation to the ModuleFile struct
* Merge ModuleFiles together into Module
* Determine child modules and load them via the same workflow

The ModuleFile initialization can be split up and handled in stages.

An alternative workflow could look like:
* An additional parameter is required that represents variable values passed in from the caller
* For each path, create a ModuleFile
  - Read the file contents
  - Parse the contents into a generic hcl structure
* Create an EvaluationContext to allow for evaluation of expressions by holding var/local values as they are discovered
* For each ModuleFile:
  - For each local/variable block
    - Validate the block
    - Add a representation to the ModuleFile struct
* Until no new variables or locals can be evaluated:
  - For each ModuleFile:
    - For each Variable in the file:
      - Look for the variable in the values passed in via the new parameter.
      - Otherwise try to evaluate the variable's default expression given the current EvaluationContext
      - If either is successful, add the discovered value to the EvaluationContext
    - For each Local in the file:
      - Try to evaluate the local's expression given the current EvaluationContext
      - If successful, add the discovered value to the EvaluationContext
  - NOTE: There is probably a better method to build a dependency tree between locals and variables to avoid the iterative try loop.
* For each ModuleFile:
  - For each non variable/local block:
    - Validate the block structure
    - Add a representation to the ModuleFile struct
* Merge ModuleFiles together into Module
* For each child module:
  - Evaluate the parameters (where possible) to the child module given the current EvaluationContext
  - Use the same workflow to load the child module, passing in the parameters that were able to be evaluated
  - Care has to be taken here to differentiate "parameter not able to be computed yet" and "parameter not specified, use the default"

There are many cases not supported by this workflow.  Any dynamically computed value will not be available during this evaluation flow.  Examples include but are not limited to: resource outputs, module outputs, data outputs.

One thing not discussed thus-far is the ability to use variables in the "root" module.  These are usually specified via tfvars, -var flags, or environment variables during the plan and apply stage.  There is a case to be made to allow these in the root module, though I am not sure how reasonable of a use case that is.

### Rationale and alternatives

There has been much discussion on this feature over nearly a decade.  Many proposal have been created and this RFC takes inspiration from many of them.

The main alternative usually proposed is to add a new constant construct to the language.  This adds clarity for what properties are guaranteed static vs potentially dynamic.  In an earlier hack and slash session in the codebase, I one variant of that [here](https://github.com/opentofu/opentofu/compare/main...cam72cam:opentofu:const_block_experiment).  It was a more straightforward change and I did appreciate the clarity of "only a constant can be used here".  Given that it is requested for the statically computed values to be able to be passed as module variables and preserve their constant nature, I felt that it was possibly too limited.  Additionally, it adds dramatically new language features that would break existing linters / code tools.  It is still an option if the current proposal is too "radical", but I think the downsides are much greater with the explicitly const approach.

### Downsides

Although syntactically identical to Terraform, this would mark a clear divergence between the feature sets of Terraform / OpenTofu configuration language.  My hope is that HashiCorp would make similar changes on their side to match our feature set, but that is not guaranteed.  It would mean that certain modules would not function if they opt-in to this new functionality.  However it is a clear error when terraform tries to parse any of these dynamic properties.

We will also need to make sure that this does not incur a performance impact for large configurations.  My initial testing has not suggested this, but it is additional complexity that will need to be considered.

### Unresolved Questions

* How should this work be broken up / divided into sub-tasks?
* ~~Should label name interpolation be considered as part of this work?~~ No
* ~~Should variables be recognized from the Environment / CLI during init for static evaluation?~~ Yes

## OpenTofu:
* https://github.com/opentofu/opentofu/issues/286
* https://github.com/opentofu/opentofu/issues/450
* https://github.com/opentofu/opentofu/issues/1017

## Terraform

* https://github.com/hashicorp/terraform/issues/13603
* https://github.com/hashicorp/terraform/issues/24187
* https://github.com/hashicorp/terraform/issues/24796
* https://github.com/hashicorp/terraform/issues/31993
  - This has been implemented in my PoC, but I am not sure if this should be included in the first pass of this feature.
* https://github.com/hashicorp/terraform/issues/32041

### Modules
* https://github.com/hashicorp/terraform/issues/1439
* https://github.com/hashicorp/terraform/issues/1846
* https://github.com/hashicorp/terraform/issues/2940
* https://github.com/hashicorp/terraform/issues/8165
* https://github.com/hashicorp/terraform/issues/14745
* https://github.com/hashicorp/terraform/issues/15613
* https://github.com/hashicorp/terraform/issues/15614
* https://github.com/hashicorp/terraform/issues/17994
* https://github.com/hashicorp/terraform/issues/18063
* https://github.com/hashicorp/terraform/issues/21891
* https://github.com/hashicorp/terraform/issues/23948
* https://github.com/hashicorp/terraform/issues/25587
* https://github.com/hashicorp/terraform/issues/26544
* https://github.com/hashicorp/terraform/issues/26619
* https://github.com/hashicorp/terraform/issues/26888
* https://github.com/hashicorp/terraform/issues/27624
* https://github.com/hashicorp/terraform/issues/28912
* https://github.com/hashicorp/terraform/issues/31701
* https://github.com/hashicorp/terraform/issues/33793

### Backends
* https://github.com/hashicorp/terraform/issues/13022
* https://github.com/hashicorp/terraform/issues/13289
* https://github.com/hashicorp/terraform/issues/17288
* https://github.com/hashicorp/terraform/issues/18632
* https://github.com/hashicorp/terraform/issues/22088
* https://github.com/hashicorp/terraform/issues/26592

### Lifecycle
* https://github.com/hashicorp/terraform/issues/3116
* https://github.com/hashicorp/terraform/issues/5346
* https://github.com/hashicorp/terraform/issues/5446
* https://github.com/hashicorp/terraform/issues/5481
* https://github.com/hashicorp/terraform/issues/10730

### Proof of Concept

This code is *incredibly* rough and implements the majority of features described in this RFC.  https://github.com/opentofu/opentofu/compare/main...cam72cam:opentofu:pre_evaluate_variables_locals.

However, some of the features (like backend configurations) were not ported from my earlier work with the const block experimentation https://github.com/opentofu/opentofu/compare/main...cam72cam:opentofu:const_block_experiment.  In the end though, the flow for those components would be nearly identical.

This only implements a subset of the features mentioned in the related issues.  However, the module versioning and backend configuration are the hardest IMO and a solid design there will allow the rest to be added over time with less effort.

**Comment by janosdebugs** [comment]
It sounds reasonable, but I'd like to point out that this can create non-obvious variable dependency issues. I'd like to add a task, however tiny, to verify that the error messages on this are incredibly clear to the user.

**Comment by marcinwyszynski** [comment]
I really like this proposal. Re:

> Should variables be recognized from the Environment / CLI during init for static evaluation?

I believe the answer is YES.

**Comment by cam72cam** [comment]
> It sounds reasonable, but I'd like to point out that this can create non-obvious variable dependency issues. I'd like to add a task, however tiny, to verify that the error messages on this are incredibly clear to the user.

Well said, it will be critical to make sure the user understands when variables / locals can't be used as constants.

**Comment by cube2222** [comment]
I think it would be good to write out how this feature will interact with other features of OpenTofu, premortem-style, what could go wrong, etc.

Additionally, it would be good to expand on how the UX for the user changes regarding the init phase. That is, will they now have to pass variables to the init phase? What if they don’t pass a variable that has a default that is used as const. What if a variable that is used as a const during init changes between init and plan phases, and how do we handle that.

---

Overall, I really like this proposal!

**Comment by cam72cam** [comment]
> will they now have to pass variables to the init phase? What if they don’t pass a variable that has a default that is used as const.
Yes that would be the largest functional difference on the CLI side.  Defaults could still be used if a variable is no passed in via -var or tfvars.  If a default is not given and it is not specified it would be considered missing and result in an error.

> What if a variable that is used as a const during init changes between init and plan phases, and how do we handle that.
We would handle it the same way we handle someone editing the TF files between init and plan.  We would provide them with an error about what changed between init/plan.  If that is insufficient, we could track the values provided as vars to the root module and add that to the error/warning.

**Comment by Yantrio** [comment]
Thanks for submitting this RFC @cam72cam !

Overall its a great proposal for a feature that people have wanted for quite some time.

My only real feedback is that I'm not much of a fan of the label interpolation mainly because I think it could cause a nightmare unless we handle renames correctly. I would propose that we leave that out of the implementation.

**Comment by cam72cam** [comment]
Thanks for the feedback @Yantrio!

I agree about the label interpolation and have had similar conversations with other folks.  I'll mark that as not-in-scope for this RFC.

**Comment by cam72cam** [comment]
I've been tinkering around with this code and working on figuring out some of the pieces that will allow better error messages.  During that, I finally looked into what _override.tf files are and how they are handled.

They allow you to merge additional configuration into your module, instead of appending.  The use cases I could find primarily talk about module sources, provider configurations, and backends.  It sounds like a nasty hack to avoid building a pre-processor.

**Comment by joeaawad** [comment]
@cam72cam I saw you put https://github.com/hashicorp/terraform/issues/3116 in the related issues section. Is that in scope for this RFC? I ask because I was going to make an RFC for resolving that issue if not, but wanted to make sure to avoid duplication if you're already considering that issue here.

**Comment by cam72cam** [comment]
@joeaawad Those are some great examples of expansions on top of this RFC and the possibilities it opens up.  My main goal for this issue is to tackle module versioning and variables in the backend configuration to try to not blow up the scope.

Once the core of this work is in place however, we can evaluate all of the additional places in the config system that could benefit from these changes.

You can certainly create new, more focused, issues for these additional cases that we can discuss and prioritize.

**Comment by ImIOImI** [comment]
I'm not 100% sure my use case is applicable or not but I use a global module to hold general config info about all my various workspaces. Would I be able to store values for my backend configuration in said module and have it inherited by multiple different templates? I'm thinking some kind of syntax like:

```
terraform {
    backend "s3" {
        region = module.config.region
    }
    required_providers {
        "aws_primary" = {
            source = "hashicorp/aws"
            region = module.config.region
        }
    }
}
```

**Comment by cam72cam** [comment]
@ImIOImI That's an interesting use case.  There's a possibility to extend this work to allow computation of effectively static module outputs.  It's something that's been thrown around during discussions on the RFC a fair bit.

My goal at the moment is to limit the scope of this initial work to locals/vars only and create followup issues to track extensions of this work.

**Comment by cam72cam** [comment]
@cube2222 suggested I check with existing IDE integration, vscode seems not to care:

![image](https://github.com/opentofu/opentofu/assets/892136/dec2d157-37ed-44a2-a1ef-7745083818d6)

Most IDEs that deal with *.tf don't do deep inspection and let you do all sort of things like this and rely on the actual CLI to do deeper validation.

**Comment by janosdebugs** [comment]
@cam72cam we could file a feature request with Jetbrains. We have a few licenses distributed among the team...

**Comment by cam72cam** [comment]
Something I hadn't fully considered is how module calls will interact with for-each.  This will need to be explored.  I believe that each.keys will be known, but values may be undefined.  Same situation with the count meta argument.

**Comment by apottere** [comment]
While getting the TF plugin updated is probably the best route, you can disable that inspection for specific modules:

![image](https://github.com/opentofu/opentofu/assets/1980504/bf27b960-4fe7-4208-adb8-7a454d757462)

You can also disable it globally, but the same inspection checks properties too so not sure if you'd want to.

The second issue is that intellij thinks the module isn't downloaded.  No idea how it performs that check or if it will work if you download with a version of tofu that supports interpolations.

**Comment by ImIOImI** [comment]
This isn't directly related, but I wanted to add another potential use case in the future related to the evaluation of constant locals that is an annoyance of mine. 

I have a map defined in a template like: 
```terraform
  private_zones = {
    akv        = "privatelink.vaultcore.azure.net",
    redis      = "privatelink.redis.cache.windows.net",
    servicebus = "privatelink.servicebus.windows.net",
    search     = "privatelink.search.windows.net"
    storage    = "privatelink.blob.core.windows.net"
  }
```

and I want to be able to pick one in a variable like:
```terraform
variable "type" {
  type        = string
  description = "Type of resource you are connecting to"
  validation {
    condition     = contains(["akv", "redis", "search", "servicebus", "storage"], var.type)
    error_message = "Allowed values for context are \"akv\", \"redis\", \"search\", \"servicebus\",  \"storage\"."
  }
}
```

So, now I've got three places to update instead of just one. It would be cool if I could do something like:

```terraform
variable "type" {
  type        = string
  description = "Type of resource you are connecting to"
  validation {
    condition     = contains(keys(local.private_zones), var.type)
    error_message = "Allowed values for context are ${formatlist(\"%s,\", keys(local.private_dns)}"
  }
}
```

**Comment by cam72cam** [comment]
That's a good use case that would be made possible by a fairly straightforward extension of this work :+1:

I'd want to double check that existing code paths won't cover that use case first.

**Comment by cam72cam** [comment]
Marking as Accepted as the TSC has met and discussed it.  I am in the process of breaking up the work directly and indirectly referred to in this RFC into manageable chunks so we can get a few people working on it.

**Comment by cam72cam** [comment]
The core of this has gone through the RFC process and has been implemented.

**Comment by ImIOImI** [comment]
<img src="https://media0.giphy.com/media/2xO491sY6f0cM/giphy.gif"/>

**Comment by sidneiweber** [comment]
I tested the alpha version and it sounds like a dream. Congratulations!

**Comment by mm-chia** [comment]
Hello,

I've tried using Tofu 1.8.0-alpha1 (from github page) and this doesn't work for me:

```
locals {
    module_version = "1.5.5"
}

module {
    source = "git@bitbucket.org:myorg/mymodule.git/?ref=v${local.module_version}"
}
```

Terraform init works, but terraform plan doesn't work.
I am getting these errors:
`│ Error: Unsuitable value type
│
│   on main.tf line 428, in module "mymodule":
│  428:   source = "git@bitbucket.org:myorg/mymodule.git/?ref=v${local.module_version}"
│
│ Unsuitable value: value must be known
╵
╷
│ Error: Variables not allowed
│
│   on main.tf line 428, in module "mymodule":
│  428:   source = "git@bitbucket.org:myorg/mymodule.git/?ref=v${local.module_version}"
│
│ Variables may not be used here.`

Anyone knows what I am doing wrong?

**Comment by davidcodesido** [comment]
I'm testing this alpha version and there's one thing I don't understand. When I define my modules with the src as a variable like: 
```
module "resource_group" {
    source = "${var.repo_url}//src/resource_group?ref=develop"
    project = terraform.workspace
    location = var.location
}
```

tofu plan works as expected and generates a plan, but when I try to apply such plan it complains that the variables changed (went to the default path for the repo)

```
│ Error: Module source has changed
│ 
│   on main.tf line 4, in module "resource_group":
│    4:     source = "${var.repo_url}//src/resource_group?ref=develop"
│ 
│ The source address was changed from
│ "git::XXXXXXXXX//src/resource_group?ref=develop"](XXXXXXX//src/resource_group?ref=develop%22)
│ to
│ "git::YYYYYYYYYYY//src/resource_group?ref=develop"
│ since this module was installed. Run "tofu init" to install all modules
│ required by this configuration.
```

 and I can't apply a var-file like I did for the plan.


```
╷
│ Error: Can't set variables when applying a saved plan
│ 
│ The -var and -var-file options cannot be used when applying a saved plan
│ file, because a saved plan includes the variable values that were set when
│ it was created.
```

Did I miss something?

**Comment by RDhar** [comment]
@davidcodesido At for the second part, that's consistent with Terraform's expected behaviour.

When applying a saved plan file, you cannot pass in a -var-file since the plan file already contains it.

**Comment by davidcodesido** [comment]
Yeah, agreed. But then there's no way to make it work if the module source is defined as variable?. Is that then a bug or a feature request for this alpha version? Running the plan, verifying it and then applying it is quite basic behavior imho, specially when running it via CD pipelines.

**Comment by RDhar** [comment]
@davidcodesido Ah sorry, I misread that - that does sound like a bug to me.

Perhaps TF has yet to bake the module/source-related variables into the plan file.

**Comment by davidcodesido** [comment]
By TF you mean Terraform or is it something Opentofu handles on its own? This looks like breaking the functionality targetted by this PR. Either you can use variables in the module sources or you can plan/apply in two steps, but not both. Unless you can think on a workaround I'm not seeing.

**Comment by ollevche** [comment]
Hey @davidcodesido! Thank you for reporting this issue. I did a quick test to reproduce the behavior, and it worked as expected: apply succeeded without needing a vars file (the values are stored inside a plan file). Could you please share the exact configuration for your scenario?

**Comment by davidcodesido** [comment]
Sure, I share all I can given this has company configurations. 

I'm running tofu 1.8.0-alpha1.

The repo_url is stored in a tfvars file:

`repo_url = "git::https://XXXXXXXXXXXX"`

And the variable is defined with a default value in the variables.tf:

```
variable "repo_url" {
    description = "The URL of the repository containing the modules"
    default = "YYYYYYYYYYYYYYY"
    type = string
} 
```
in my test I'm running two stages of an azure pipeline (two different VMs are spawned in azure for that purpose). 

First I run the  init command in both stages, same command in both passing the var file:

`tofu init -backend-config="resource_group_name=$(tfresourcegroup)" -backend-config="subscription_id=$(tfsubscription)" -backend-config="storage_account_name=$(tfstorageaccount)" -backend-config="container_name=$(tfcontainername)" -backend-config="key=$(tfstatekey)" -backend-config="access_key=$(tfaccesskey)" -var-file=$(tfvarsfile.secureFilePath)`

I have a workspace which I select/create if needed with :

`tofu workspace select workspace1 || tofu workspace new workspace1`

And last I create a plan in the first stage with:

`tofu plan -out=tfplan -var-file=$(tfvarsfile.secureFilePath)`

Then this plan is stored and passed to the other stage(VM) which then runs the apply with:

`tofu apply -auto-approve tfplan`

On the second stage I make sure the plan is there by running:

`cat tfplan`

Which displays the binary content, however when I try a tofu show on that plan:

`tofu show -json tfplan`

It already displays the same error as the apply:

```
Error: Couldn't show local plan
│ 
│ Plan read error: 2 problems:
│ 
│ - Module source has changed: The source address was changed from
│ "git::[https://XXXXXXXXXX//src/keyvault?ref=develop"](https://XXXXXXXXXX//src/keyvault?ref=develop%22)
│ to
│ "git::ssh://YYYYYYYYYY/***-modules//src/keyvault?ref=develop"
│ since this module was installed. Run "tofu init" to install all modules
│ required by this configuration.
│ - Module source has changed: The source address was changed from
│ "git::[https://XXXXXXXXXX//src/resource_group?ref=develop"](https://XXXXXXXXXX/_git/***-modules//src/resource_group?ref=develop%22)
│ to
│ "git::ssh://YYYYYYYYYY/***-modules//src/resource_group?ref=develop"
│ since this module was installed. Run "tofu init" to install all modules
│ required by this configuration.
```

**Comment by ollevche** [comment]
> I'm running tofu 1.8.0-alpha1.

Hey @davidcodesido! Please, upgrade to a stable version, we fixed a couple of things since `alpha1`. This particular issue was addressed in https://github.com/opentofu/opentofu/pull/1771.

**Comment by davidcodesido** [comment]
Works like a charm. Apologies for the noise, wasn't aware that this made it into a stable version yet. Thanks for your help.

**Comment by ollevche** [comment]
No problem, we are here to help! You can find more info about the 1.8.0 GA release in [the blogpost](https://opentofu.org/blog/opentofu-1-8-0/).

**Comment by RDhar** [comment]
> @davidcodesido:
> Running the plan, verifying it and then applying it is quite basic behavior imho, specially when running it via CD pipelines.

Totally agreed, including the 2-part provisioning of TF (either Tofu or Terraform!) within automated CI/CD pipelines - in my case, using GitHub Actions instead of Azure.

In order to test and verify the early-evaluation of variables within modules, I've published a public Action pipeline using the latest stable v1.8.0 Tofu on [this PR](https://github.com/DevSecTop/TF-via-PR/pull/266), where the module's source version is [specified via variable](https://github.com/DevSecTop/TF-via-PR/blob/2caec58c257584d3afbeb45bac17a4e885e6577c/sample/modules/s3_bucket/main.tf#L2-L4).

```tf
# sample/modules/s3_bucket/main.tf
module "s3_bucket" {
  source  = "terraform-aws-modules/s3-bucket/aws"
  version = var.source_version
```

Super-neat to see `init`, `workspace`, `fmt`, `plan` and `apply` all working together as expected, and even raising helpful errors when I didn't pass in the module version variable correctly at first.

Better yet, we can see the workflow logs first-hand for each of the separate [`plan`](https://github.com/DevSecTop/TF-via-PR/actions/runs/10181079448/job/28160370248#step:5:115) (which stores the plan file as an artifact) and [`apply`](https://github.com/DevSecTop/TF-via-PR/actions/runs/10181113950/job/28160485353#step:5:115) (which retrieves the latest plan file) workflow runs.

Exciting stuff, thanks @ollevche and @cam72cam!
